aws/code/workflows/MLWorkflowMultiGPU.py

```diff
@@ -0,0 +1,218 @@
+    def fit ( self, X_train, y_train ):       
+        """ Fit decision tree model [ architecture defined by HPO parameters ] """
+        if 'XGBoost' in self.hpo_config.model_type:
+            print('> fit xgboost model')
+            dtrain = xgboost.dask.DaskDMatrix( self.client, X_train, y_train)
+            xgboost_output = xgboost.dask.train( self.client, self.hpo_config.model_params, dtrain, 
+                                                num_boost_round = self.hpo_config.model_params['num_boost_round'] )
+            trained_model = xgboost_output['booster']
+
+        elif 'RandomForest' in self.hpo_config.model_type:
+            print('> fit randomforest model')
+            trained_model = RandomForestClassifier ( n_estimators = self.hpo_config.model_params['n_estimators'],
+                                                     max_depth = self.hpo_config.model_params['max_depth'],
+                                                     max_features = self.hpo_config.model_params['max_features'],
+                                                     n_bins = self.hpo_config.model_params['n_bins'] )\
+                                                     .fit( X_train, y_train.astype('int32') )                                                     
+        return trained_model 
+    
+    @timer_decorator
+    def predict ( self, trained_model, X_test, threshold = 0.5 ):
+        """ Inference with the trained model on the unseen test data """
+
+        print('> predict with trained model ')
+        if 'XGBoost' in self.hpo_config.model_type:
+            dtest = xgboost.dask.DaskDMatrix( self.client, X_test )
+            predictions = xgboost.dask.predict( self.client, trained_model, dtest).compute() 
+            predictions = (predictions > threshold ) * 1.0                    
+            
+        elif 'RandomForest' in self.hpo_config.model_type:
+            predictions = trained_model.predict( X_test ).compute()
+
+        return predictions
+
+    @timer_decorator
+    def score ( self, y_test, predictions ): 
+        """ Score predictions vs ground truth labels on test data """
+        print('> score predictions')
+        y_test = y_test.compute()
+        score = accuracy_score ( y_test.astype( self.hpo_config.dataset_dtype ),
+                                predictions.astype( self.hpo_config.dataset_dtype ) )
+
+        print(f'\t score = {score}')
+        self.cv_fold_scores += [score]
+        return score
+
+    def save_best_model ( self, score, trained_model, filename = 'saved_model' ): 
+        """  Persist/save model that sets a new high score """
+
+        if score > self.best_score:
+            self.best_score = score
+            print('> saving high-scoring model')
+            output_filename = self.hpo_config.model_store_directory + '/' + str( filename )
+            if 'XGBoost' in self.hpo_config.model_type:
+                trained_model.save_model( output_filename + '_mgpu_xgb' )
```

Comment:
Prefer `.format`, f-strings, or `join` to `+` operator for concatenating strings